cocotb/types/__init__.py

```diff
@@ -0,0 +1,43 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+from .range import Range  # noqa: F401
+from .array import Array  # noqa: F401
+
+
+def concat(a: Array, b: Array) -> Array:
+    """
+    Create a new array that is the concatenation of one array with another.
+
+    Uses the :meth:`__concat__` or :meth:`__rconcat__` special methods to dispatch to a particular implementation,
+    exactly like other binary operations in Python.
+
+    Raises:
+        TypeError: when the arguments do not support concatenation in the given order.
+    """
+    a_concat = getattr(a.__class__, "__concat__", None)
+    a_rconcat = getattr(a.__class__, "__rconcat__", None)
+    b_rconcat = getattr(b.__class__, "__rconcat__", None)
```

Comment:
Strictly the semantics should be:
```suggestion
    a_concat = getattr(type(a), "__concat__", None)
    a_rconcat = getattr(type(a), "__rconcat__", None)
    b_rconcat = getattr(type(b), "__rconcat__", None)
```

Reply:
Is there a difference between `type` and `__class__` in Python 3? I have read that they are equivalent, and `__class__` is faster. I have used `__class__` everywhere in these PRs, if there is a good reason to switch I guess I should take this as a general comment.

Reply:
```
In [90]: class Foo:
    ...:     @property
    ...:     def __class__(self):
    ...:         return "oops"
    ...:

In [91]:

In [91]: type(Foo())
Out[91]: __main__.Foo

In [92]: Foo().__class__
Out[92]: 'oops'
```

Reply:
More realistically, this comes up if someone defines an object with a custom `__getattribute__`

Reply:
Alright. I'll update all the occurrences in this PR.